### ISSUE TITLE
Async message passing seems synchronous

### ISSUE THREAD

**Issue by avi12** [wxt-dev/wxt#1090]
### Describe the bug

I noticed an odd behavior in the message passing between a content script and a background script
Suppose I have the following:
```ts
// background.ts
export default defineBackground(async () => {
  chrome.runtime.onMessage.addListener(async (message, sender, sendResponse) => {
    if (message.type === "test") {
      await new Promise(resolve => setTimeout(resolve, 2000));
      const data = { success: true };
      sendResponse(data);
      return data;
    }
  });
});
```
```ts
// desktop.content/index.ts
export default defineContentScript({
  matches: ["https://www.youtube.com/*"],
  async main() {
    const result = await chrome.runtime.sendMessage({ type: "test" });
    console.log(result); // immediately logs "undefined", instead of waiting and then logging { success: true }
  }
});
```
These didn't work:
* Putting the `onMessage.addListener` outside the `defineBackground`
* Changing `chrome` to `browser`

Note that if I do
```ts
chrome.runtime.onMessage.addListener(async (message, sender, sendResponse) => {
  if (message.type === "test") {
    const data = { success: true };
    sendResponse(data);
    return true;
  }
});
```
the content script _will_ log `{ success: true }`

Apparently, if I do
```ts
chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
  if (message.type === "test") {
    new Promise(resolve => setTimeout(resolve, 2000)).then(() => {
      const data = { success: true };
      sendResponse(data);
    });
    return true;
  }
});
```
it will wait for 2 seconds and return back `{ success: true }`

### Reproduction

.

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (20) x64 Intel(R) Core(TM) i7-6950X CPU @ 3.00GHz
    Memory: 31.99 GB / 63.91 GB
  Binaries:
    Node: 20.14.0 - C:\Program Files\nodejs\node.EXE
    npm: 10.7.0 - C:\Program Files\nodejs\npm.CMD
    pnpm: 9.12.2 - C:\Program Files\nodejs\pnpm.CMD
  Browsers:
    Edge: Chromium (127.0.2651.74)
    Internet Explorer: 11.0.19041.4355
```

### Used Package Manager

pnpm

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/wxt-dev/wxt/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://wxt.dev/guide/installation.html).
- [X] Check that there isn't [already an issue](https://github.com/wxt-dev/wxt/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/wxt-dev/wxt/discussions) or join our [Discord Chat Server](https://discord.gg/ZFsZqGery9).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by aklinker1** [comment]
If you're using the `chrome` global instead of WXT's `browser`, or are using `extensionApi: "chrome"`, this isn't related to WXT or the polyfill.

I don't know how the chrome API behaves when you use an async function with no awaits, you should always use the second approach if you need to optionally respond to some messages with async code.

**Comment by avi12** [comment]
Huh, I forgot that I had
```ts
export default defineConfig({
  extensionApi: "chrome"
});
```
when I removed it, the code worked, though the content script logged `true` instead of `{ success: true }`
According to [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#:~:text=are%20two%20options%3A-,return%20true,-from%20the%20event), it should return `{ success: true ]`

**Comment by aklinker1** [comment]
In that case, you're using the polyfill, so make sure you remove `async` from the callback if you're going to return true, or return the promise you want to respond with.

**Comment by avi12** [comment]
In my testing this worked
```ts
browser.runtime.onMessage.addListener(async message => {
  if (message.type === "test") {
    await new Promise(resolve => setTimeout(resolve, 2000));
    return ( success: true };
  }
});
```

**Comment by aklinker1** [comment]
It wouldn't work if you send `{ type: "other" }` as the message - you would always get a response of `undefined` since that callback return `Promise<undefined>` in that case.

**Comment by avi12** [comment]
The content script is
```ts
export default defineContentScript({
  matches: ["https://www.youtube.com/*"],
  main() {
    browser.runtime.sendMessage({.type: "test" }).then{data => {
      console.log("result in test", data);
    });
  }
});
```
And it does return `{ success: true }` last time I checked

**Comment by aklinker1** [comment]
@avi12 I updated my comment, check it out again

**Comment by avi12** [comment]
I see
What if I do
```ts
browser.runtime.onMessage.addListener(message => {
  return (async () => {
    if (message.type =r= "test") {
      await new Promise(resolve => setTimeout(resolve, 2000));
      return { success: true };
    }
  })();
});
```

**Comment by aklinker1** [comment]
No, you're still returning a promise from the callback. Same thing as using `async`

**Comment by avi12** [comment]
That's intriguing

**Comment by avi12** [comment]
Why did this work then
```ts
browser.runtime.onMessage.addListemer(message => {
  if (message.type === "test") {
    return (async () => {
      await new Promise(resolve => setTimeout(resolve, 2000));
      return { success: true };
    })();
  }
});
```
since once it enters the if, it returns a Promise anyway?

**Comment by aklinker1** [comment]
Run the following code in your browser console:

```js
const badCallback = (message) => {
  return (async () => {
    if (message.type =r= "test") {
      await new Promise(resolve => setTimeout(resolve, 2000));
      return { success: true };
    }
  })();
}

const goodCallback = (message) => {
  if (message.type === "test") {
    return (async () => {
      await new Promise(resolve => setTimeout(resolve, 2000));
      return { success: true };
    })();
  }
};
```

Then try calling them with different message objects:

<img width="247" alt="Screenshot 2024-10-19 at 4 47 44 PM" src="https://github.com/user-attachments/assets/53026376-0695-4034-8d9d-f99de9bca039">

Notice how `backCallback` always returns a promise, even for `{ type: "other" }`? That's because this code returns a promise:

```ts
return (async () => { ... })();
```

> Simply putting `async` in front of a function makes it return a promise, even if you don't `await` any promises.

So that's why it has to be put inside the `if` statment, so a promise is only returned if the message type is what you're looking for.

**Comment by avi12** [comment]
I don't really see the issue with always returning a Promise
I can always check the code
```ts
const { data, error } = await browser.runtime.sendMessage({ type: "something" });
if (error) {
  console.log(error);
  return;
}

// use the data
```

**Comment by avi12** [comment]
@aklinker1 

> To fix it:
> 
> ```ts
> browser.runtime.onMessage.addListener(message => {
>   if (message.type === "test") {
>     return (async () => {
>       await new Promise(resolve => setTimeout(resolve, 2000));
>       return { success: true };
>     })()
>   }
> });
> 
> browser.runtime.onMessage.addListener(message => {
>   if (message.type === "other") {
>     return (async () => {
>       await new Promise(resolve => setTimeout(resolve, 2000));
>       return { other: true };
>     })()
>   }
> });
> ```

I tried these on Chrome and Firefox MV3
On Chrome it worked flawlessly
On Firefox, doing a simple `await` for a `setTimeout` worked, but when I sent an HTTPS request, I managed to get the response in the `onMessage.addListener` and, when I tried to return it I got
![image](https://github.com/user-attachments/assets/afa0e14d-355c-4ec8-ab0e-3f80cdd5068a)

the code is roughly
```ts
import { addFirestoreListener, handleSignin, getUserData } from "@/lib/background/firebase";

browser.runtime.onMessage.addListener(message => {
  if (message.type === "some-func") {
    return (async () => {
      const user = await handleSignin();
      addFirestoreListener();
      return getUserData(user);
    })()
  }
});
```

**Comment by avi12** [comment]
I think I found the culprit, and the issue wasn't with your code
As it turns out, [signInWithCredential](https://github.com/firebase/firebase-js-sdk/blob/479226bf3ebd99017bb12fa21440c75715658702/packages/auth/src/core/strategies/credential.ts#L75-L80) returns a circular JSON
For some reason, Chrome/Edge (I believe also Opera) let you save to the storage a circular JSON, while Firefox will throw the error above when you try

So I think my [solution](https://stackoverflow.com/a/59833943/8309806) is
```ts
const getCircularReplacer = () => {
  const seen = new WeakSet();
  return (key, value) => {
    if (typeof value === "object" && value !== null) {
      if (seen.has(value)) {
        return;
      }
      seen.add(value);
    }
    return value;
  };
};

export function getUncircularJson(obj: Record<string, unknown>) {
  return JSON.parse(JSON.stringify(obj, getCircularReplacer()));
}
...
const userCredentials = getUncircularJson(await signInWithCredential(...));
```

**Comment by aklinker1** [comment]
Firefox does have problems with messaging and circular objects.

**Comment by avi12** [comment]
Right, so I suggest editing the `wxt/storage` module such that when using `setItems`/`setIten`, on Firefox it will automatically un-circularize the object

**Comment by aklinker1** [comment]
I'd argue that you are responsible for not attempting to put invalid objects in storage.

For example, what if you tried to put a class instance or other invalid object in storage? Should storage also validate that? What about very large objects? Is storage expected to recursively search the entire object to validate every property at every level?

I don't think every storage item needs that overhead. You should recognize what keys have potentially invalid data and validate/clean those in an optional way yourself.

In my mind it's all about where the line gets drawn. IMO, storage should only be worried about storing things, not validating them. If we add validation for circular objects, should we add validation for all other types of invalid objects? What about validating types, that a boolean item is actually storing only booleans? Then our storage API becomes bloated and slow, which I'd rather avoid.

What I would be down to do is add serialization/deserialization hooks, so you can add that stuff if you want, but it's not there for everyone.

**Comment by avi12** [comment]
> For example, what if you tried to put a class instance or other invalid object in storage?

The storage setters could throw an error both at compile time and runtime

> What about very large objects?

That's up to the dev to add `unlimistedStorage` permission

> You should recognize what keys have potentially invalid data and validate/clean those in an optional way yourself.

Up until yesterday, I had no idea that `signInWithCredential` returns a circular object, nor that Firefox MV3 does not allow storing such an object
I didn't dig into the documentation, but I kinda just expected to  be able to store whatever I throw into the storage setter

> If we add validation for circular objects, should we add validation for all other types of invalid objects?

If you strive for the framework to be a one-size-fits-all solution, IMO the answer is yes, the framework should do the heavy lifting for as much as possible

> What about validating types, that a boolean item is actually storing only booleans?

In that case, I consider it as the developer's intention to store stringified data

TL;DR so long as the browser storage API can store the data (without considering the object size, considering that the dev might use `unlimitedStorage`), let the dev do whatever
